dulwich/client.py

```diff
@@ -1102,16 +1107,50 @@ def __init__(self, hostname):
 class SubprocessSSHVendor(SSHVendor):
     """SSH vendor that shells out to the local 'ssh' command."""
 
-    def run_command(self, host, command, username=None, port=None):
-        # FIXME: This has no way to deal with passwords..
-        args = ['ssh', '-x']
-        if port is not None:
-            args.extend(['-p', str(port)])
-        if username is not None:
+    def run_command(self, host, command, username=None, port=None,
+                    password=None, key_filename=None):
+        if password and key_filename:
+            raise NotImplementedError(
+                "You can't set passphrase for ssh key with "
+                "SubprocessSSHVendor, use ParamikoSSHVendor instead"
+            )
+
+        if sys.platform == 'win32':
+            args = ['putty.exe', '-ssh']
+            if password:
+                args.extend(['-pw', str(password)])
+        else:
+            if password:
+                try:
+                    subprocess.call(["sshpass"])
```

Comment:
I'm uncomfortable with supporting sshpass like this; I'm sure there are some legitimate uses for it, but on e.g. a multi-user system this will cause the password to show up in the process list.

Reply:
Problem that there is no good multi platform way for obtain the same with just only subprocess. If I right remember, for sending password directly into stdin of ssh we should use fork or threads or asinc solution and most of all default solutions do it. We can use `pexpect` for it, but again, on windows this solution require a [special way](https://pexpect.readthedocs.io/en/stable/overview.html#windows) and it possibly would have a conflict with current `dulwitch` wrapper

In my opinion `paramiko` variant would be the best option for security and amount of features in compare with subprocess variant. But it you don't want to support this, and user already have `sshpass` (or `putty` in case ow windows) and this user don't want to use the key auth or preconfigured ssh config file instead - security is not a problem of `dulwitch` itself...

Reply:
see https://github.com/dulwich/dulwich/issues/364. The paramiko module doesn't have any tests and my experience in another project is that its API isn't super stable, which is why I moved it to contrib.

If somebody were to contribute end-to-end tests for paramiko, i.e. tests that setup a SSH server and verify that running dulwich using paramiko works over it, I'd be happy to move it back to dulwich proper (and e.g. making it the default on Windows).

Reply:
Okay :)
Is there some ready-to-go tests for the subprocess based class that we can use for it?
There is a [paramiko test for client](https://github.com/paramiko/paramiko/blob/master/tests/test_client.py) under LGPL that we can use, also there are a lot of some mock ssh servers and even docker image with git server and ssh.
But imo we can do a easier\better way to test ssh channel: we can add a read only ssh key into this repo (it called 'deployment key') and test all operation over real existed data :)
We can use it for both native ssh client and paramiko.

We can lock paramiko dependency into one latest specific version and check it properly as other solution

Reply:
Replying on #364, to keep these two conversations separate.